File: frontend/lib/household.ts
/**
 * Build PolicyEngine household situations for WATCA calculations.
 * TypeScript port of watca_calc/household.py
 */

import { HouseholdRequest } from "./types";

const GROUP_UNITS = ["families", "spm_units", "tax_units", "households"] as const;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function addMemberToUnits(situation: Record<string, any>, memberId: string) {
  for (const unit of GROUP_UNITS) {
    const key = Object.keys(situation[unit])[0];
    situation[unit][key].members.push(memberId);
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function buildHouseholdSituation(params: HouseholdRequest): Record<string, any> {
  const { age_head, age_spouse, dependent_ages, year, max_earnings, state_code } = params;
  const yearStr = String(year);

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const situation: Record<string, any> = {
    people: { you: { age: { [yearStr]: age_head } } },
    families: { "your family": { members: ["you"] } },
    marital_units: { "your marital unit": { members: ["you"] } },
    spm_units: { "your household": { members: ["you"] } },
    tax_units: { "your tax unit": { members: ["you"], adjusted_gross_income: { [yearStr]: null } } },
    households: {
      "your household": {
        members: ["you"],
        state_code: { [yearStr]: state_code },
        household_net_income: { [yearStr]: null },
      },
    },
  };

  // Add AGI sweep axis
  situation.axes = [
    [
      {
        name: "adjusted_gross_income",
        min: 0,
        max: max_earnings,
        count: Math.min(4001, Math.max(501, Math.floor(max_earnings / 500))),
        period: yearStr,
        target: "tax_unit",
      },
    ],
  ];

  // Add spouse
  if (age_spouse !== null && age_spouse !== undefined) {
    situation.people["your partner"] = { age: { [yearStr]: age_spouse } };
    addMemberToUnits(situation, "your partner");
    situation.marital_units["your marital unit"].members.push("your partner");
  }

  // Add dependents
  for (let i = 0; i < dependent_ages.length; i++) {
    let childId: string;
    if (i === 0) {
      childId = "your first dependent";
    } else if (i === 1) {
      childId = "your second dependent";
    } else {
      childId = `dependent_${i + 1}`;
    }

    situation.people[childId] = { age: { [yearStr]: dependent_ages[i] } };
    addMemberToUnits(situation, childId);
    situation.marital_units[`${childId}'s marital unit`] = {
      members: [childId],
    };
  }

  return situation;
}

File: watca_calc/household.py
"""Build PolicyEngine household situations for WATCA calculations."""

_GROUP_UNITS = ["families", "spm_units", "tax_units", "households"]


def _add_member_to_units(situation, member_id):
    """Append a member to all group units (families, spm, tax, households)."""
    for unit in _GROUP_UNITS:
        key = next(iter(situation[unit]))
        situation[unit][key]["members"].append(member_id)


def build_household_situation(
    age_head: int,
    age_spouse: int | None,
    dependent_ages: list[int],
    year: int = 2026,
    with_axes: bool = False,
    max_earnings: int = 2_000_000,
    state_code: str = "CA",
) -> dict:
    """
    Build a PolicyEngine household situation dict.

    Args:
        age_head: Age of household head.
        age_spouse: Age of spouse (None if single).
        dependent_ages: List of dependent ages.
        year: Tax year.
        with_axes: If True, adds AGI sweep axis.
        max_earnings: Maximum AGI for the sweep axis.

    Returns:
        PolicyEngine situation dict.
    """
    situation = {
        "people": {"you": {"age": {year: age_head}}},
        "families": {"your family": {"members": ["you"]}},
        "marital_units": {"your marital unit": {"members": ["you"]}},
        "spm_units": {"your household": {"members": ["you"]}},
        "tax_units": {"your tax unit": {"members": ["you"]}},
        "households": {
            "your household": {
                "members": ["you"],
                "state_code": {year: state_code},
            }
        },
    }

    if with_axes:
        situation["axes"] = [
            [
                {
                    "name": "adjusted_gross_income",
                    "min": 0,
                    "max": max_earnings,
                    "count": min(4_001, max(501, max_earnings // 500)),
                    "period": year,
                    "target": "tax_unit",
                }
            ]
        ]

    if age_spouse is not None:
        situation["people"]["your partner"] = {
            "age": {year: age_spouse}
        }
        _add_member_to_units(situation, "your partner")
        situation["marital_units"]["your marital unit"][
            "members"
        ].append("your partner")

    for i, dep_age in enumerate(dependent_ages):
        if i == 0:
            child_id = "your first dependent"
        elif i == 1:
            child_id = "your second dependent"
        else:
            child_id = f"dependent_{i + 1}"

        situation["people"][child_id] = {"age": {year: dep_age}}
        _add_member_to_units(situation, child_id)
        situation["marital_units"][f"{child_id}'s marital unit"] = {
            "members": [child_id]
        }

    return situation
